File: apps/backend/app/core/telemetry.py
import uuid
import logging
from typing import Dict, List, Any, Optional
from datetime import datetime

logger = logging.getLogger("telemetry")

class TelemetryStore:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(TelemetryStore, cls).__new__(cls)
            cls._instance.events: List[Dict[str, Any]] = []
        return cls._instance

    def record(
        self,
        run_id: str,
        metric: str,
        value: float,
        agent_type: Optional[str] = None,
        tool_name: Optional[str] = None
    ) -> None:
        event = {
            "id": str(uuid.uuid4()),
            "run_id": run_id,
            "metric": metric,
            "value": value,
            "agent_type": agent_type,
            "tool_name": tool_name,
            "timestamp": datetime.utcnow().isoformat()
        }
        self.events.append(event)
        logger.debug(f"Telemetry recorded: {metric}={value} for agent={agent_type}")

    def get_events_for_run(self, run_id: str) -> List[Dict[str, Any]]:
        return [e for e in self.events if e["run_id"] == run_id]

telemetry = TelemetryStore()

File: packages/agents/src/telemetry/telemetry.ts
import type { TelemetryEvent, AgentType } from "@testpilot/types";
import { createLogger } from "@testpilot/shared";

const logger = createLogger("telemetry");

/**
 * Dedicated Observability and Telemetry Engine.
 * Tracks metrics: reasoning latency, tool latency, token consumption, retries, and failure rates.
 */
export class TelemetryStore {
  private static instance: TelemetryStore;
  private events: TelemetryEvent[] = [];

  private constructor() {}

  public static getInstance(): TelemetryStore {
    if (!TelemetryStore.instance) {
      TelemetryStore.instance = new TelemetryStore();
    }
    return TelemetryStore.instance;
  }

  /** Record a metric event */
  public record(
    runId: string,
    metric: TelemetryEvent["metric"],
    value: number,
    agentType?: AgentType,
    toolName?: string
  ): void {
    const event: TelemetryEvent = {
      id: crypto.randomUUID(),
      runId,
      metric,
      value,
      agentType,
      toolName,
      timestamp: new Date(),
    };

    this.events.push(event);
    logger.debug({ runId, metric, value, agentType, toolName }, "Telemetry event recorded");
  }

  /** Get telemetry events for a specific run */
  public getEventsForRun(runId: string): TelemetryEvent[] {
    return this.events.filter((e) => e.runId === runId);
  }

  /** Calculate total tokens used for a run */
  public getTotalTokens(runId: string): number {
    return this.events
      .filter((e) => e.runId === runId && e.metric === "tokens_used")
      .reduce((sum, e) => sum + e.value, 0);
  }
}

export const telemetry = TelemetryStore.getInstance();
